core/lib/types/src/kzg.rs

```diff
@@ -0,0 +1,315 @@
+    pub fn new(pubdata: Vec<u8>) -> Self {
+        let kzg_settings = Self::kzg_settings();
+
+        assert!(pubdata.len() <= BYTES_PER_BLOB_ZK_SYNC);
+
+        let mut zksync_blob = [0u8; BYTES_PER_BLOB_ZK_SYNC];
+        zksync_blob[0..pubdata.len()].copy_from_slice(&pubdata);
+
+        let mut hasher = Keccak256::new();
+        hasher.update(zksync_blob);
+        let linear_hash = &hasher.finalize_reset();
+
+        let bytes_4844 = zksync_pubdata_into_ethereum_4844_data(&zksync_blob);
+        let blob = Blob::new(bytes_4844.try_into().unwrap());
+
+        let kzg_commitment = KzgCommitment::blob_to_kzg_commitment(&blob, &kzg_settings).unwrap();
+
+        let mut hasher = Keccak256::new();
+        hasher.update(kzg_commitment.to_bytes().into_inner());
+        let mut versioned_hash_bytes = hasher.finalize_reset();
+        versioned_hash_bytes[0] = 0x01;
+        let versioned_hash = Bytes32::from_bytes(&versioned_hash_bytes).unwrap();
+
+        hasher.update(linear_hash.as_slice());
+        hasher.update(versioned_hash_bytes);
+
+        let opening_point_bytes = &hasher.finalize_reset();
+        let mut opening_point = [0u8; 32];
+        opening_point[16..].copy_from_slice(&opening_point_bytes[16..]);
+        let opening_point = Bytes32::from_bytes(&opening_point).unwrap();
+
+        let (opening_proof, opening_value) =
+            KzgProof::compute_kzg_proof(&blob, &opening_point, &kzg_settings).unwrap();
+
+        let blob_proof =
+            KzgProof::compute_blob_kzg_proof(&blob, &kzg_commitment.to_bytes(), &kzg_settings)
+                .unwrap();
+
+        Self {
+            blob,
+            kzg_commitment,
+            opening_point,
+            opening_value,
+            opening_proof,
+            versioned_hash,
+            blob_proof,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use serde::{Deserialize, Serialize};
+    use serde_with::serde_as;
+
+    use super::{KzgInfo, KzgProof};
+    use crate::{H256, U256};
+
+    #[serde_as]
+    #[derive(Debug, Serialize, Deserialize)]
+    struct ExepectedOutputs {
+        versioned_hash: H256,
+        #[serde_as(as = "serde_with::hex::Hex")]
+        kzg_commitment: Vec<u8>,
+        opening_point: U256,
+        opening_value: U256,
+        #[serde_as(as = "serde_with::hex::Hex")]
+        opening_proof: Vec<u8>,
+        #[serde_as(as = "serde_with::hex::Hex")]
+        blob_proof: Vec<u8>,
+        #[serde_as(as = "serde_with::hex::Hex")]
+        pubdata_commitment: Vec<u8>,
+    }
+
+    #[serde_as]
+    #[derive(Debug, Serialize, Deserialize)]
+    struct KzgTest {
+        #[serde_as(as = "serde_with::hex::Hex")]
+        pubdata: Vec<u8>,
+        expected_outputs: ExepectedOutputs,
+    }
+
+    #[test]
+    fn kzg_test() {
+        let zksync_home = std::env::var("ZKSYNC_HOME").unwrap_or_else(|_| ".".into());
+        let path = std::path::Path::new(&zksync_home).join("etc/kzg_tests/kzg_test_0.json");
```

Comment:
how was this test generated?

Reply:
generated by hand but ill use a known working test case

Reply:
i ended up generating a test case:
heres the blob txn: https://goerli.etherscan.io/tx/0x18a931affb89d2378b53834a4d791ec0eb781f2c0f55968e9696ab2e9283222d

beacon explorer showing the side car (blob 5): https://goerli.beaconcha.in/block/10400517#blobs

getting data of blob: https://dora.goerli.ethpandaops.io/slot/7431299

note that the data sent is after processing with `zksync_pubdata_into_ethereum_4844_data`